File: app/src/main/java/com/raghif/app/data/QueueLogic.kt
package com.raghif.app.data

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

// Pure queue logic, extracted from OwnerQueueScreen so the admin state machine is
// unit-testable. Statuses: waiting (bought, not called) -> notified (batch notif sent,
// awaiting arrival) -> collected (admin checked them in).
const val STATUS_WAITING = "waiting"
const val STATUS_NOTIFIED = "notified"
const val STATUS_COLLECTED = "collected"

/** Admin arrival check: notified (pending) <-> collected (arrived). Waiting rows have no action. */
fun toggleArrivalStatus(status: String): String = when (status) {
    STATUS_NOTIFIED -> STATUS_COLLECTED
    STATUS_COLLECTED -> STATUS_NOTIFIED
    else -> status
}

/** Smallest batch number that still has waiting customers — the one the admin should notify next. */
fun nextBatchToNotify(queue: List<PurchaseEntity>): Int? =
    queue.filter { it.status == STATUS_WAITING }.minByOrNull { it.batchNumber }?.batchNumber

/** Admin search: matches name, phone, or national ID (رقم الهوية). Blank query returns the queue as-is. */
fun filterQueue(
    queue: List<PurchaseEntity>,
    usersByUserId: Map<String, UserEntity>,
    query: String
): List<PurchaseEntity> {
    val q = query.trim()
    if (q.isEmpty()) return queue
    return queue.filter { p ->
        val u = usersByUserId[p.userId]
        u?.phone?.contains(q) == true ||
            u?.name?.contains(q, ignoreCase = true) == true ||
            u?.personalId?.contains(q) == true
    }
}

const val BATCH_INTERVAL_MINUTES = 10

/** Estimated ready time (epoch millis): createdAt + batch x interval — batch is that bakery's queue position. */
fun estimatedReadyAtMillis(createdAt: Long, batchNumber: Int): Long =
    createdAt + batchNumber * BATCH_INTERVAL_MINUTES * 60_000L

private val READY_TIME_FORMAT = DateTimeFormatter.ofPattern("HH:mm")

/** Formats epoch millis as HH:mm in the device's local time. */
fun formatReadyTime(millis: Long): String =
    Instant.ofEpochMilli(millis).atZone(ZoneId.systemDefault()).format(READY_TIME_FORMAT)

/** True when the admin must set today's allocation first: it's a new day, or the current bundles are gone. */
fun needsAllocation(store: StoreEntity, today: String): Boolean =
    store.allocationDate != today || store.bagsRemaining <= 0

File: flutter/lib/features/queue/queue_logic.dart
import '../../domain/models/purchase_model.dart';
import '../../domain/models/store_model.dart';

/// Pure queue logic, kept separate from widgets so it's unit-testable without
/// pumping a widget tree. Mirrors app/.../data/QueueLogic.kt's rules.
const batchIntervalMinutes = 10;

String todayDateString([DateTime? now]) {
  final d = now ?? DateTime.now();
  final y = d.year.toString().padLeft(4, '0');
  final m = d.month.toString().padLeft(2, '0');
  final day = d.day.toString().padLeft(2, '0');
  return '$y-$m-$day';
}

/// Admin arrival check: notified (pending) <-> collected (arrived). Waiting
/// rows have no action.
PurchaseStatus toggleArrivalStatus(PurchaseStatus status) => switch (status) {
  PurchaseStatus.notified => PurchaseStatus.collected,
  PurchaseStatus.collected => PurchaseStatus.notified,
  PurchaseStatus.waiting => PurchaseStatus.waiting,
};

/// Smallest batch number that still has waiting customers — the one the
/// owner should notify next. Null when nobody is waiting.
int? nextBatchToNotify(List<PurchaseModel> queue) {
  final waiting = queue.where((p) => p.status == PurchaseStatus.waiting);
  if (waiting.isEmpty) return null;
  return waiting.map((p) => p.batchNumber).reduce((a, b) => a < b ? a : b);
}

/// Estimated ready time (epoch millis): createdAt + batch x interval.
int estimatedReadyAtMillis(int createdAtMillis, int batchNumber) =>
    createdAtMillis + batchNumber * batchIntervalMinutes * 60000;

/// Formats epoch millis as HH:mm in local time.
String formatReadyTime(int millis) {
  final dt = DateTime.fromMillisecondsSinceEpoch(millis).toLocal();
  final h = dt.hour.toString().padLeft(2, '0');
  final m = dt.minute.toString().padLeft(2, '0');
  return '$h:$m';
}

/// True when the owner must set today's allocation first: it's a new day, or
/// the current bundles are gone.
bool needsAllocation(StoreModel store, String today) =>
    store.allocationDate != today || store.bagsRemaining <= 0;

/// In-app low-stock warning threshold. Actual SMS/push delivery is out of
/// scope for this prototype (decision D) — the app only surfaces the state.
const int lowStockThreshold = 20;

/// True when the store is running low and the owner should be warned in-app.
bool isLowStock(StoreModel store) => store.bagsRemaining <= lowStockThreshold;

/// Today's paid-but-not-collected buyers — the dashboard's "still waiting"
/// count, so the owner doesn't have to open the queue to know.
int pendingPickupCount(List<PurchaseModel> queue) =>
    queue.where((p) => p.status != PurchaseStatus.collected).length;

/// Tallies for one batch header in the owner's queue.
class BatchProgress {
  const BatchProgress({
    required this.batch,
    required this.total,
    required this.picked,
    required this.awaitingPickup,
    required this.waiting,
  });

  final int batch;
  final int total;

  /// Handed over.
  final int picked;

  /// Called, but the buyer hasn't shown up yet.
  final int awaitingPickup;

  /// Not called yet.
  final int waiting;

  bool get allPicked => picked == total;
}

/// Per-batch tallies for [queue], ordered by batch number — "called N / picked
/// M / still waiting K" next to each batch header.
List<BatchProgress> batchProgressForQueue(List<PurchaseModel> queue) {
  final byBatch = <int, List<PurchaseModel>>{};
  for (final purchase in queue) {
    byBatch.putIfAbsent(purchase.batchNumber, () => []).add(purchase);
  }
  final batches = byBatch.keys.toList()..sort();
  return [
    for (final batch in batches)
      BatchProgress(
        batch: batch,
        total: byBatch[batch]!.length,
        picked: byBatch[batch]!
            .where((p) => p.status == PurchaseStatus.collected)
            .length,
        awaitingPickup: byBatch[batch]!
            .where((p) => p.status == PurchaseStatus.notified)
            .length,
        waiting: byBatch[batch]!
            .where((p) => p.status == PurchaseStatus.waiting)
            .length,
      ),
  ];
}

/// Buyers in *earlier* batches who were called but never picked up. Releasing
/// a new batch on top of them is what the owner gets warned about.
int outstandingBefore(int batch, List<PurchaseModel> queue) => queue
    .where(
      (p) =>
          p.batchNumber < batch && p.status == PurchaseStatus.notified,
    )
    .length;
